profiler/internal/cmemprof/pprof_test.go

```diff
@@ -0,0 +1,61 @@
+//	  whitespace.
+//
+// The following /proc/self/maps example was obtained by compiling a C program
+// which just calls "sleep(1000)", giving it a name with a space ("space daemon"),
+// and deleting the executable after starting the program.
+
+var procMaps = []byte(`55f747992000-55f747993000 r--p 00000000 08:01 788178                     /usr/bin/space daemon (deleted)
+55f747993000-55f747994000 r-xp 00001000 08:01 788178                     /usr/bin/space daemon (deleted)
+55f747994000-55f747995000 r--p 00002000 08:01 788178                     /usr/bin/space daemon (deleted)
+55f747995000-55f747996000 r--p 00002000 08:01 788178                     /usr/bin/space daemon (deleted)
+55f747996000-55f747997000 rw-p 00003000 08:01 788178                     /usr/bin/space daemon (deleted)
+7fa3b3c00000-7fa3b3c22000 r--p 00000000 08:01 3463                       /usr/lib/x86_64-linux-gnu/libc-2.31.so
+7fa3b3c22000-7fa3b3d9a000 r-xp 00022000 08:01 3463                       /usr/lib/x86_64-linux-gnu/libc-2.31.so
+7fa3b3d9a000-7fa3b3de8000 r--p 0019a000 08:01 3463                       /usr/lib/x86_64-linux-gnu/libc-2.31.so
+7fa3b3de8000-7fa3b3dec000 r--p 001e7000 08:01 3463                       /usr/lib/x86_64-linux-gnu/libc-2.31.so
+7fa3b3dec000-7fa3b3dee000 rw-p 001eb000 08:01 3463                       /usr/lib/x86_64-linux-gnu/libc-2.31.so
+7fa3b3dee000-7fa3b3df4000 rw-p 00000000 00:00 0 
+7fa3b3dfc000-7fa3b3dfd000 r--p 00000000 08:01 3459                       /usr/lib/x86_64-linux-gnu/ld-2.31.so
+7fa3b3dfd000-7fa3b3e20000 r-xp 00001000 08:01 3459                       /usr/lib/x86_64-linux-gnu/ld-2.31.so
+7fa3b3e20000-7fa3b3e28000 r--p 00024000 08:01 3459                       /usr/lib/x86_64-linux-gnu/ld-2.31.so
+7fa3b3e29000-7fa3b3e2a000 r--p 0002c000 08:01 3459                       /usr/lib/x86_64-linux-gnu/ld-2.31.so
+7fa3b3e2a000-7fa3b3e2b000 rw-p 0002d000 08:01 3459                       /usr/lib/x86_64-linux-gnu/ld-2.31.so
+7fa3b3e2b000-7fa3b3e2c000 rw-p 00000000 00:00 0 
+7fff1e028000-7fff1e049000 rw-p 00000000 00:00 0                          [stack]
+7fff1e04c000-7fff1e04f000 r--p 00000000 00:00 0                          [vvar]
+7fff1e04f000-7fff1e050000 r-xp 00000000 00:00 0                          [vdso]
+ffffffffff600000-ffffffffff601000 --xp 00000000 00:00 0                  [vsyscall]
+`)
+
+func Test_parseMapping(t *testing.T) {
+	want := []*profile.Mapping{
+		{ID: 1, Start: 0x55f747993000, Limit: 0x55f747994000, Offset: 0x00001000, File: "/usr/bin/space daemon"},
+		{ID: 2, Start: 0x7fa3b3c22000, Limit: 0x7fa3b3d9a000, Offset: 0x00022000, File: "/usr/lib/x86_64-linux-gnu/libc-2.31.so"},
+		{ID: 3, Start: 0x7fa3b3dfd000, Limit: 0x7fa3b3e20000, Offset: 0x00001000, File: "/usr/lib/x86_64-linux-gnu/ld-2.31.so"},
+		{ID: 4, Start: 0x7fff1e04f000, Limit: 0x7fff1e050000, Offset: 0x00000000, File: "[vdso]"},
+		{ID: 5, Start: 0xffffffffff600000, Limit: 0xffffffffff601000, Offset: 0x00000000, File: "[vsyscall]"},
+	}
```

Comment:
NIT: Might be nice to intentionally change `procMaps` so that mappings are not listed in `Start` order so that the `sort.Slice` behavior gets tested.

Reply:
That wouldn't hurt. FWIW, every /proc/self/maps I've looked at has its entries already sorted by starting address. In fact, the [implementation](https://sourcegraph.com/github.com/torvalds/linux@0fac198def2b41138850867b6aa92044c76ff802/-/blob/fs/proc/task_mmu.c?L351-361) traverses a linked list of virtual memory areas which are sorted by address ([ref](https://sourcegraph.com/github.com/torvalds/linux@0fac198def2b41138850867b6aa92044c76ff802/-/blob/include/linux/mm_types.h?L398-406)). The proc man page doesn't make any promises about the display ordering, though.

Also, the official [pprof proto file](https://github.com/google/pprof/blob/a41b82acbcb1f6a6d0998b0fddeb1f1adb7d8fc1/proto/profile.proto#L58-L60) says that the first mapping should be the main executable. In practice I always see the executable as the first mapping, and I think by convention the executable (text section, bss, etc) is loaded into lower addresses and then libraries, heap, etc. are mapped to higher addresses. I don't know how you'd tell which mapping is the main executable otherwise. Maybe look for an entry that isn't called `[vdso]` or `[vsyscall]` and doesn't end in `.so`... and hope there's only one?

Reply:
I'm okay with leaving the code as is. The motivation behind this NIT was just to ensure the invariant needed for the binary search lookup of the mappings.